tests/unit/test_geneset.py

```diff
@@ -0,0 +1,98 @@
+        )
+
+        # test all geneset attributes
+        self.assertCountEqual(gs.genes, ["FHIT", "MYC"])
+        self.assertEqual(len(gs._genes), 2)
+        self.assertEqual(gs.source, "UCSC")
+        self.assertEqual(gs.species, "Homo sapiens")
+
+    def test_empty_set(self):
+        # test creating empty geneset
+        gs = Geneset(MagicMock())
+        self.assertCountEqual(gs.genes, [])
+
+    def test_dont_override_genes(self):
+        gs = Geneset(
+            MagicMock(), genes=["FHIT", "MYC"], source="UCSC", species="Homo sapiens"
+        )
+        gs2 = Geneset(MagicMock(), genes=["BRCA2", "MYC", "ABC"])
+        gs | gs2
+
+        # the original genesets should not change when performing set operations on them
+        self.assertCountEqual(gs.genes, ["FHIT", "MYC"])
+        self.assertCountEqual(gs2.genes, ["BRCA2", "MYC", "ABC"])
+
+    def test_set_operations(self):
+        gs = Geneset(
+            MagicMock(), genes=["FHIT", "MYC"], source="UCSC", species="Homo sapiens"
+        )
+        gs2 = Geneset(MagicMock(), genes=["BRCA2", "MYC", "ABC"])
+        gs_union = gs | gs2
+        gs_intersection = gs & gs2
+        gs_difference = gs - gs2
+        gs_rsub = gs.__rsub__(gs2)
+        gs_symmetric_difference = gs ^ gs2
+
+        # test set operations
+        self.assertCountEqual(gs_union.genes, ["FHIT", "MYC", "BRCA2", "ABC"])
+        self.assertCountEqual(gs_intersection.genes, ["MYC"])
+        self.assertCountEqual(gs_difference.genes, ["FHIT"])
+        self.assertCountEqual(gs_rsub.genes, ["BRCA2", "ABC"])
+        self.assertCountEqual(gs_symmetric_difference.genes, ["FHIT", "BRCA2", "ABC"])
+
+    def test_fetch_data(self):
+        gs = Geneset(
+            MagicMock(),
+            output={"geneset_json": 1, "source": "UCSC", "species": "Homo sapiens"},
+            id=1,
+        )
+        response = MagicMock()
+        response.content = '{"json": {"genes": ["FHIT", "MYC"]}}'.encode("utf-8")
+        gs.resolwe = MagicMock(**{"session.get.return_value": response, "url": ""})
+
+        # test all geneset attributes
+        self.assertCountEqual(gs.genes, ["FHIT", "MYC"])
+        self.assertEqual(gs.source, "UCSC")
+        self.assertEqual(gs.species, "Homo sapiens")
+
+    @patch("resdk.resources.geneset.Geneset", spec=True)
+    def test_update_fields(self, base_mock):
+        base_mock.configure_mock(
+            id=1, slug="test", _original_values={"slug": "test-old"}
+        )
+        base_mock.WRITABLE_FIELDS = ("slug",)
+
+        base_mock.api = MagicMock(**{"patch.return_value": {"id": 1, "slug": "test"}})
+        Geneset.save(base_mock)
+        self.assertEqual(base_mock._update_fields.call_count, 1)
+
+    def test_create(self):
+        gs = Geneset(
+            MagicMock(), genes=["FHIT", "MYC"], source="UCSC", species="Homo sapiens"
+        )
+        gs.api = MagicMock(**{"post.return_value": {"id": 1, "slug": "test"}})
+        gs.save()

```

Comment:
These two lines can actually be deleted and test will still pass :-)
I know that it can be challenging to write a good test with so much mocking. Consider just checking that post request was made with expected payload.